Allow Embedded Cache sizes to be configured via the CLI

ryanemerson opened the pull request:
Closes #31514

I have excluded the `work` cache as I don't believe a user should ever be able to define an upper-bound on this, as it could cause issues with events if miss-configured. I have also not exposed the `*Revisions` caches, as they are not described to users in the caching guide and their max-count is already set based upon the configured value of the associated cache.

ryanemerson committed:
Allow Embedded Cache sizes to be configured via the CLI

Closes #31514

Signed-off-by: Ryan Emerson <remerson@redhat.com>

ryanemerson commented:
@keycloak/cloud-native I based my approach to this PR on https://github.com/keycloak/keycloak/pull/26756, however I noticed that there isn't any tests present in that PR to verify that the specified CLI values are applied to the cache configuration. Is there an existing set of test(s) that I can re-use/take inspiration from to verify that the max-count is in fact configured as expected when the values are present?

shawkins commented:
To test just the options / mappers you can use [ConfigurationTest](https://github.com/keycloak/keycloak/blob/44321b3d6c92b8f7ec8db75b552ae7053a8cf8d0/quarkus/runtime/src/test/java/org/keycloak/quarkus/runtime/configuration/test/ConfigurationTest.java#L49) or create another subclass. For end-to-end testing there are the Dist tests https://github.com/keycloak/keycloak/tree/main/quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist

ryanemerson commented:
Thanks for the pointer @shawkins. I have added a test to `ConfigurationTest` and also added `ClusterConfigKeepAliveDistTest`. I tried adding a new dist test to `ClusterConfigDistTest`, however I was unable to call the new resource I expose when `keepAlive=false`, hence why the new class was created. Maybe I'm doing something wrong?

pruivo commented on quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/CachingPropertyMappers.java:
```diff
@@ -21,61 +24,71 @@ private CachingPropertyMappers() {
+        PropertyMapper<?>[] maxCountMappers = Stream.of(CachingOptions.MAX_COUNT_OPTIONS)
+              .map(opt ->
+                    fromOption(opt)
+                          .isEnabled(InfinispanUtils::isEmbeddedInfinispan)
```
local caches option should always be enabled. 
Probably it is cleaner to add 2 new arrays in `InfinispanConnectionProvider`
* local bounded caches -> always available to be configured
* clustered bounded caches -> only when in "embedded mode"

ryanemerson replied:
Currently the `keycloak-config-api` module doesn't have a dependency on `keycloak-model-infinispan` and has minimal dependencies in general, so I'm not sure if we can add this?

pruivo replied:
oh :( having the arrays in CacheOption should be fine.

pruivo commented on quarkus/config-api/src/main/java/org/keycloak/config/CachingOptions.java:
```diff
@@ -26,6 +27,17 @@ public class CachingOptions {
     public static final String CACHE_METRICS_HISTOGRAMS_ENABLED_PROPERTY = CACHE_METRICS_PREFIX + "-histograms-enabled";
 
+    public static final String[] MAX_COUNT_CACHES = new String[]{"authorization", "keys", "realms", "users", "actionTokens",
+          "authenticationSessions", "clientSessions", "loginFailures", "offlineSessions", "offlineClientSessions", "sessions"};
```
`actionToken`, `authenticationSessions`, and `loginFailures` data are not stored in the database. Those cannot be bounded.

pruivo reviewed (changes requested):
A couple of things. 
* The CLI does not support negative values:
```
kc.sh start-dev --cache=ispn --cache-embedded-sessions-max-count=-1
Missing required value. Option '--cache-embedded-sessions-max-count' (max-count) expects a single value.
```
* hmm... I'm not sure how to say this, but eviction is never enabled. We never configured `when-full` in the XML or programmatically.

ryanemerson commented on quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:
```diff
@@ -313,6 +315,7 @@ private CompletableFuture<DefaultCacheManager> startEmbeddedCacheManager(String
                 configureRemoteStores(builder);
             }
             configureSessionsCaches(builder);
+            configureCacheMaxCount(builder);
```
I have updated `configureCacheMaxCount` so it takes a `String[]` of cache names, we can then call with the local cache names for both embedded and remote.

ryanemerson commented:
> The CLI does not support negative values:

Is that an issue? My understanding was that a negative value means that there's no upper bound on the number of entries in the cache, which is the same as not defining a max-count.

shawkins commented:
> Is that an issue? My understanding was that a negative value means that there's no upper bound on the number of entries in the cache, which is the same as not defining a max-count.

It is an issue: https://github.com/keycloak/keycloak/issues/33068

I'll see if there is an easy fix.

pruivo commented:
> > The CLI does not support negative values:
> 
> Is that an issue? My understanding was that a negative value means that there's no upper bound on the number of entries in the cache, which is the same as not defining a max-count.

`max-count` is enabled by default in "all" caches. I'm just thinking of using these CLI options to disable it, but we don't have to.

pruivo reviewed (approved):
LGTM, thanks @ryanemerson

ryanemerson commented:
> `max-count` is enabled by default in "all" caches. I'm just thinking of using these CLI options to disable it, but we don't have to.

I see, thanks for explaining. That would be a nice option to have if/when #33068 is resolved.

vmuzikar commented on docs/guides/server/caching.adoc:
It'd be nice to also have a mention of this in the release notes.

vmuzikar commented on quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/CachingPropertyMappers.java:
```diff
@@ -128,4 +146,11 @@ private static String getDefaultTruststorePathValue() {
+
+    private static PropertyMapper<?> maxCountOpt(String cacheName, BooleanSupplier isEnabled) {
+        return fromOption(CachingOptions.maxCountOption(cacheName))
+              .isEnabled(isEnabled)
```
It'd also good to have the `enabledWhen` message set here.

vmuzikar commented on quarkus/tests/integration/src/test-providers/java/org/keycloak/it/resource/realm/TestRealmResource.java:
```diff
@@ -17,20 +17,33 @@
+
+    final InfinispanConnectionProvider infinispanConnectionProvider;
+
+    public TestRealmResource(InfinispanConnectionProvider infinispanConnectionProvider) {
```
To be more future proof, I'd rather suggest passing the whole `KeycloakSession` here instead of a the specific provider.

vmuzikar reviewed (commented):
@ryanemerson Thank you for the PR!

vmuzikar commented:
One more question. When #33068 is resolved, should we also document how `max-count` behaves with `-1`? If so, it'd be good to have a follow-up issue created for it.

ryanemerson commented:
> One more question. When #33068 is resolved, should we also document how `max-count` behaves with `-1`? If so, it'd be good to have a follow-up issue created for it.

Good idea, I have created #33146

vmuzikar commented:
@ryanemerson Do you need the `-1` functionality in 26.0, or can it wait for 26.1?

ryanemerson commented:
It would be nice to have, but I don't think it's a blocker \cc @pruivo

pruivo commented:
@vmuzikar @ryanemerson not a blocker for 26.0. 
A closer look at the Infinispan code, we may not need support for negative values. Zero seems to create an unbounded cache. 

https://github.com/infinispan/infinispan/blob/d6a64da6ef608894dfbb4b9047920a62c755028f/core/src/main/java/org/infinispan/configuration/cache/MemoryConfiguration.java#L163

mabartos reviewed (commented):
@ryanemerson Great work! LGTM from the Quarkus perspective.

Just one thing: It seems the CI failing due to the `HelpCommandDistTest`. Please, build the Quarkus distribution, and then execute the test with the env specified here: https://github.com/keycloak/keycloak/blob/main/quarkus/README.md#updating-expectations

It will automatically update the approved files.

ryanemerson commented:
Thanks for the pointer @mabartos, I've fixed the HelpCommandDistTest files and squashed the commits.

mabartos reviewed (approved):
@ryanemerson Thanks, LGTM.

ahus1 reviewed (approved):
Approving based on previous reviews. Thank you to everyone involved in this!
